service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/auth/TokenCredentialCache.java

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.sdk.iot.service.auth;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.credential.TokenRequestContext;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+
+/**
+ * This class generates AAD authentication tokens from a TokenCredential but caches previous tokens when they aren't near
+ * expiry.
+ */
+public class TokenCredentialCache
+{
+    private final static int MINUTES_BEFORE_PROACTIVE_RENEWAL = 9;
```

Comment:
I made it a round figure of 10 on .net :D But 9 is fine too.

Reply:
I did a "less than or equal to" operation, so our functions should work the same in including 9 minutes